tests/src/Integration/Action/SystemPageRedirectTest.php

```diff
@@ -0,0 +1,144 @@
+
+    /**
+     * The mocked logger.
+     *
+     * @var \Psr\Log\LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * The action to be tested.
+     *
+     * @var \Drupal\rules\Plugin\RulesAction\SystemPageRedirect
+     */
+    protected $action;
+
+    /**
+     * The mocked request stack service.
+     *
+     * @var \Symfony\Component\HttpFoundation\RequestStack
+     */
+    protected $request_stack;
+
+    /**
+     * A mocked request.
+     *
+     * @var \Symfony\Component\HttpFoundation\Request
+     */
+    protected $current_request;
+
+    /**
+     * A mocked parameter bag.
+     *
+     * @var \Symfony\Component\HttpFoundation\ParameterBag
+     */
+    protected $parameter_bag;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setUp() {
+      parent::setUp();
+
+      $this->logger = $this->prophesize(LoggerInterface::class);
+      $this->container->set('logger.factory', $this->logger->reveal());
+
+      $this->parameter_bag = $this->prophesize(ParameterBag::class);
+
+      $this->current_request = $this->prophesize(Request::class);
+      $this->current_request->attributes = $this->parameter_bag->reveal();
+
+      $this->request_stack = $this->prophesize(RequestStack::class);
+      $this->request_stack->getCurrentRequest()->willReturn($this->current_request);
+      $this->container->set('request_stack', $this->request_stack->reveal());
+
+      $this->action = $this->actionManager->createInstance('rules_page_redirect');
+    }
+
+    /**
+     * Tests redirection.
+     *
+     * @covers ::execute
+     */
+    public function testRedirect() {
+      $this->current_request->getPathInfo()->willReturn('/some/random/test/path');
+
+      $this->action->setContextValue('url', '/test/url');
+      $this->action->execute();
+
+      $this->parameter_bag->set('rulesRedirectActionUrl', '/test/url')->shouldHaveBeenCalled();
+    }
+
+    /**
+     * Tests unsuccessful redirection due to ongoing batch process.
+     *
+     * @covers ::execute
+     */
+    public function testRedirectBatch() {
+      $this->current_request->getPathInfo()->willReturn('/some/random/test/path');
+      batch_set('Test batch!');
+
+      $this->action->setContextValue('url', '/test/url');
+      $this->action->execute();
+
+      $this->logger->log(LogLevel::WARNING, 'Skipped page redirect during batch processing.')->shouldHaveBeenCalled();
+    }
+
+    /**
+     * Tests unsuccessful redirection due to rules admin page location.
+     *
+     * @covers ::execute
+     */
+    public function testRedirectRulesAdminPage() {
+      $this->current_request->getPathInfo()->willReturn('/admin/config/workflow/rules');
+
+      $this->action->setContextValue('url', '/test/url');
+      $this->action->execute();
+
+      $this->logger->log(LogLevel::WARNING, 'Skipped page redirect on a rules admin page.')->shouldHaveBeenCalled();
+    }
+
+  }
+}
+
+namespace {
+
+  if (!function_exists('batch_get')) {
+
```

Comment:
can we have a function comment here that we are mocking the original batch*() functions?